### ISSUE TITLE
Logs truncated

### ISSUE THREAD

**Issue by yaroslavvb** [wandb/client#368]
The training log in this run is truncated halfway.
https://app.wandb.ai/yaroslavvb/nccl_multiversion/runs/xf2u0a1t/logs

However, in my terminal things look fine
```
#
#                                                     out-of-place                       in-place          
#       size         count    type   redop     time   algbw   busbw  error     time   algbw   busbw  error
#        (B)    (elements)                     (us)  (GB/s)  (GB/s)            (us)  (GB/s)  (GB/s)       
ip-172-31-3-63:38231:38231 [0] NCCL INFO Launch mode Parallel
           8             2   float     sum    147.8    
wandb: Waiting for W&B process to finish, PID 38094
0.00    0.00  2e-07    145.9    0.00    0.00  1e-07
          16             4   float     sum    146.7    0.00    0.00  0e+00    148.7    0.00    0.00  1e-07
          32             8   float     sum    145.4    0.00    0.00  1e-07    145.2    0.00    0.00  1e-07
          64            16   float     sum    148.3    0.00    0.00  1e-07    146.2    0.00    0.00  6e-08
         128            32   float     sum    153.3    0.00    0.00  6e-08    145.7    0.00    0.00  6e-08
         256            64   float     sum    147.0    0.00    0.00  6e-08    145.0    0.00    0.00  6e-08
         512           128   float     sum    147.6    0.00    0.01  6e-08    147.5    0.00    0.01  6e-08
        1024           256   float     sum    153.0    0.01    0.01  2e-07    151.6    0.01    0.01  2e-07
        2048           512   float     sum    155.2    0.01    0.02  2e-07    154.3    0.01    0.02  2e-07
        4096          1024   float     sum    156.9    0.03    0.05  5e-07    156.1    0.03    0.05  5e-07
        8192          2048   float     sum    157.9    0.05    0.10  5e-07    159.3    0.05    0.10  5e-07
       16384          4096   float     sum    160.9    0.10    0.19  5e-07    160.3    0.10    0.19  5e-07
       32768          8192   float     sum    171.6    0.19    0.36  5e-07    169.6    0.19    0.36  5e-07
       65536         16384   float     sum    190.2    0.34    0.65  5e-07    187.6    0.35    0.66  5e-07
      131072         32768   float     sum    220.1    0.60    1.12  5e-07    219.1    0.60    1.12  5e-07
      262144         65536   float     sum    493.8    0.53    1.00  5e-07    481.1    0.54    1.02  5e-07
      524288        131072   float     sum    627.1    0.84    1.57  5e-07    623.3    0.84    1.58  5e-07
     1048576        262144   float     sum   1179.2    0.89    1.67  5e-07   1198.1    0.88    1.64  5e-07
     2097152        524288   float     sum   1880.3    1.12    2.09  5e-07   1864.5    1.12    2.11  5e-07
     4194304       1048576   float     sum   2748.1    1.53    2.86  5wandb: Program ended successfully.
e-07   2695.8    1.56    2.92  5e-07
     8388608       2097152   float     sum   4628.4    1.81    3.40  5e-07   4606.3    1.82    3.41  5e-07
    16777216       4194304   float     sum   8202.9    2.05    3.83  5e-07   8201.6    2.05    3.84  5e-07
    33554432       8388608   float     sum    15366    2.18    4.09  5e-07    15431    2.17    4.08  5e-07
    67108864      16777216   float     sum    25315    2.65    4.97  5e-07    25285    2.65    4.98  5e-07
   134217728      33554432   float     sum    50201    2.67    5.01  5e-07    50372    2.66    5.00  5e-07
   268435456      67108864   float     sum   100369    2.67    5.01  5e-07   100327    2.68    5.02  5e-07
   536870912     134217728   float     sum   200402    2.68    5.02  5e-07   200552    2.68    5.02  5e-07
  1073741824     268435456   float     sum   401202    2.68    5.02  5e-07   400422    2.68    5.03  5e-07
# Out of bounds values : 0 OK
# Avg bus bandwidth    : 1.71812 
#

wandb: Run summary:
wandb:           NCCL_HOME /home/ec2-user/nccl/...
wandb:    NCCL_VERSION_TAG 2.4.6
wandb:            _runtime 84.59366464614868
wandb:          _timestamp 1559929024.8924124
wandb:             SIZE_MB 1024
wandb:            NUM_GPUS 16
wandb:           NPER_NODE 8
wandb:           CUDA_HOME /usr/local/cuda-10.0...
wandb:           ofi_patch False
wandb:            EFA_HOME /opt/amazon/efa
wandb:              do_efa 1
wandb:         FOLDER_ROOT /home/ec2-user/nccl/...
wandb:               _step 1
wandb:            MPI_HOME /home/ec2-user/anaco...
wandb:   avg_bus_bandwidth 1.71812
wandb: Syncing files in wandb/run-20190607_173547-xf2u0a1t:
wandb:   description.md
wandb:   diff.patch
wandb:   output.log
wandb: plus 5 W&B file(s) and 0 media file(s)
wandb: - 5 of 8 files, 37277.000 of 41750.000 bytes uploaded
wandb: \ 6 of 8 files, 37315.000 of 41750.000 bytes uploaded
wandb: | 6 of 8 files, 37315.000 of 41750.000 bytes uploaded
wandb: / 6 of 8 files, 37315.000 of 41750.000 bytes uploaded
wandb: - 6 of 8 files, 37315.000 of 41750.000 bytes uploaded
wandb: \ 6 of 8 files, 37315.000 of 41750.000 bytes uploaded
wandb: | 6 of 8 files, 37315.000 of 41750.000 bytes uploaded
wandb:                                                                                
wandb: Synced bench-16-efa-stock: https://app.wandb.ai/yaroslavvb/nccl_multiversion/runs/xf2u0a1t

```

**Comment by yaroslavvb** [comment]
Here's the code that was used to log the run above https://github.com/cybertronai/aws-network-benchmarks/blob/b89e25b098c33094faa48541e5dde17e154327d7/nccl_multiversion.py#L251

**Comment by yaroslavvb** [comment]
The thing to note is that the value "1.718" shows up, this could be a buffering issue, will try with  PYTHONUNBUFFERED=1 for future runs

**Comment by lukas** [comment]
Really appreciate the bug report.  I was taking a quick look at your reference code and I noticed you have your AWS keys in there - I think you probably want to take those out as the code is public.

**Comment by yaroslavvb** [comment]
AWS makes it really easy to leak credentials, don't they. Those are fake keys though :)

**Comment by yaroslavvb** [comment]
BTW, I noticed in my console logs it says " 2.86  5wandb: Program ended successfully", and this is exactly when the web logs cut off.

**Comment by yaroslavvb** [comment]
One possibility is that this has something to do with doing multiple runs in a row, I was doing run/crash/run/crash/run/etc , ie what triggers "wandb: Program ended successfully"?

**Comment by vanpelt** [comment]
That’s triggered when your script exits, we add an exit hook to your script
that tells the wandb process how you’ve exited.  I thought we flush the
stdout buffer, but it looks like we don’t.  Can you try calling
sys.stdout.flush right before your process exits?  It’s also possible work
is happening in threads so those need to be joined before the parent
exits...

**Comment by yaroslavvb** [comment]
Will try that. Normally you don't need a flush because default behavior flushes after each print, but a possibility was that `os.dup2` modified buffering behavior of stdout stream. I needed output of the command inside inside the script, so I was doing the following.

The solution on wandb side would be to flush stdout/stderr streams before exiting

```
class capture_stdout:
  """Utility to capture output, use as follows
     with util.capture_stdout() as stdout:
        sess = tf.Session()
    print("Captured:", stdout.getvalue()).
    """

  def __init__(self, fd=STDOUT):
    self.fd = fd
    self.prevfd = None

  def __enter__(self):
    t = tempfile.NamedTemporaryFile()
    self.prevfd = os.dup(self.fd)
    os.dup2(t.fileno(), self.fd)
    return TemporaryFileHelper(t)

 def __exit__(self, exc_type, exc_value, traceback):
   os.dup2(self.prevfd, self.fd)
   print("Running command:")

print(args.internal_cmd)
with util.capture_stdout() as outfile:
    os.system(args.internal_cmd)

# print logs so they show up in console
out: str = outfile.getvalue()
print(out)
```

**Comment by vanpelt** [comment]
We may have fixed this in the most recent release 0.8.2.  If you're still seeing it would you mind making a simple script that exhibits the lost logs?

**Comment by yaroslavvb** [comment]
seems to be fixed for me as well
